components/script/dom/console.rs

```diff
@@ -55,14 +64,138 @@ where
     f()
 }
 
-fn console_messages(global: &GlobalScope, messages: &[DOMString], level: LogLevel) {
-    console_message(global, DOMString::from(messages.join(" ")), level)
+const MAX_LOG_DEPTH: usize = 10;
+
+#[allow(unsafe_code)]
+unsafe fn value_to_source(cx: *mut jsapi::JSContext, value: HandleValue) -> DOMString {
+    rooted!(in(cx) let mut js_string = std::ptr::null_mut::<jsapi::JSString>());
+    js_string.set(JS_ValueToSource(cx, value));
+    return jsstring_to_str(cx, *js_string);
+}
+
+#[allow(unsafe_code)]
+fn stringify_message(message: HandleValue) -> DOMString {
+    let cx = *GlobalScope::get_cx();
+    unsafe {
+        if message.is_string() {
+            jsstring_to_str(cx, message.to_string())
+        } else {
+            unsafe fn stringify_inner(
+                cx: *mut jsapi::JSContext,
+                value: HandleValue,
+                mut parents: Vec<u64>,
```

Comment:
A pushing and popping approach with a mutable vector would probably be a bit more efficient.

Reply:
I tried making it use a mutable vector, but it made the stringifiation and memory allocation logic quite a bit more complicated - I don't think the extra complexity is worth some slight performance gains.